src/java/picard/analysis/directed/GcBiasMetricsCollector.java

```diff
@@ -0,0 +1,286 @@
+            this.readGroup = readGroup;
+            final String prefix;
+            if (this.readGroup != null) {
+                prefix = this.readGroup;
+                gcData.put(prefix, new GcObject());
+            } else if (this.library != null) {
+                prefix = this.library;
+                gcData.put(prefix, new GcObject());
+            } else if (this.sample != null) {
+                prefix = this.sample;
+                gcData.put(prefix, new GcObject());
+            } else {
+                prefix = "All_Reads";
+                gcData.put(prefix, new GcObject());
+            }
+        }
+
+        /////////////////////////////////////////////////////////////////////////////
+        //Takes each record and sends them to addRead to calculate gc metrics for
+        // that read for each accumulation level
+        /////////////////////////////////////////////////////////////////////////////
+        public void acceptRecord(final GcBiasCollectorArgs args) {
+            final SAMRecord rec = args.getRec();
+            final String type;
+            if(!rec.getReadUnmappedFlag()) {
+                final ReferenceSequence ref = args.getRef();
+                final byte[] refBases = ref.getBases();
+                final String refName = ref.getName();
+                final byte[] gc = gcByRef.get(refName);
+                final String group;
+                if (this.readGroup != null) {
+                    type = this.readGroup;
+                    group = "Read Group";
+                    addRead(gcData.get(type), rec, group, gc, refBases);
+                } else if (this.library != null) {
+                    type = this.library;
+                    group = "Library";
+                    addRead(gcData.get(type), rec, group, gc, refBases);
+                } else if (this.sample != null) {
+                    type = this.sample;
+                    group = "Sample";
+                    addRead(gcData.get(type), rec, group, gc, refBases);
+                } else {
+                    type = "All_Reads";
+                    group = "All Reads";
+                    addRead(gcData.get(type), rec, group, gc, refBases);
+                }
+            }
+            else {
+                for (final Map.Entry<String, GcObject> entry : gcData.entrySet()) {
+                    final GcObject gcCur = entry.getValue();
+                    if (!rec.getReadPairedFlag() || rec.getFirstOfPairFlag()) ++gcCur.totalClusters;
+                }
+            }
+        }
+
+        public void finish() {}
+
+        /////////////////////////////////////////////////////////////////////////////
+        // Sums the values in an int[].
+        /////////////////////////////////////////////////////////////////////////////
+        private double sum(final int[] values) {
+            final int length = values.length;
+            double total = 0;
+            for (int i = 0; i < length; ++i) {
```

Comment:
Wow I didn't even notice that, that's old code from the original but I will change it. I've never seen ++i before!